contracts/external/cw-payroll/src/linking.rs

```diff
@@ -0,0 +1,190 @@
+        initiator_id: StreamId,
+        link_id: StreamId,
+        deps: DepsMut,
+        info: &MessageInfo,
+    ) -> Result<Response, ContractError>;
+    fn detach(
+        &self,
+        initiator_id: StreamId,
+        link_id: StreamId,
+        deps: DepsMut,
+        info: &MessageInfo,
+    ) -> Result<Response, ContractError>;
+    fn create_link_delete_msg(
+        &self,
+        stream_id: StreamId,
+        env: &Env,
+        deps: DepsMut,
+        info: &MessageInfo,
+    ) -> Result<CosmosMsg, ContractError>;
+    fn create_sync_msg(
+        &self,
+        stream_id: StreamId,
+        env: &Env,
+        deps: DepsMut,
+        info: &MessageInfo,
+        sync_type: LinkSyncType,
+    ) -> Result<CosmosMsg, ContractError>;
+}
+
+impl SupportsLinking for Stream {
+    fn link(
+        &mut self,
+        initiator_id: StreamId,
+        link_id: StreamId,
+        deps: DepsMut,
+        info: &MessageInfo,
+    ) -> Result<Response, ContractError> {
+        // let mut initiator = STREAMS.may_load(deps.storage, initiator_id)?.ok_or(
+        //     ContractError::InitiatorStreamNotFound {
+        //         stream_id: initiator_id,
+        //     },
+        // )?;
+
+        let mut link = STREAMS
+            .may_load(deps.storage, initiator_id)?
+            .ok_or(ContractError::StreamNotFound {})?;
+
+        if self.admin != info.sender {
+            return Err(ContractError::Unauthorized {});
+        }
+        self.link_id = Some(link_id);
+        self.is_link_initiator = true;
+        link.link_id = Some(initiator_id);
+        link.is_link_initiator = false;
+
+        save_stream(deps.storage, initiator_id, self).unwrap();
+        save_stream(deps.storage, link_id, &link).unwrap();
+
+        let response = Response::new()
+            .add_attribute("method", "link")
+            .add_attribute("initiator_id", initiator_id.to_string())
+            .add_attribute("link_id", link_id.to_string())
+            .add_attribute("admin", info.sender.clone());
+
+        Ok(response)
+    }
+
+    fn detach(
+        &self,
+        initiator_id: StreamId,
+        link_id: StreamId,
+        deps: DepsMut,
+        info: &MessageInfo,
+    ) -> Result<Response, ContractError> {
+        let mut initiator = STREAMS.may_load(deps.storage, initiator_id)?.ok_or(
+            ContractError::InitiatorStreamNotFound {
+                stream_id: initiator_id,
+            },
+        )?;
+
+        if !initiator.is_detachable {
+            return Err(ContractError::StreamNotDetachable {});
+        }
+
+        let mut link = STREAMS
+            .may_load(deps.storage, initiator_id)?
+            .ok_or(ContractError::StreamNotFound {})?;
+
+        if initiator.admin != info.sender {
+            return Err(ContractError::Unauthorized {});
+        }
+
+        initiator.link_id = None;
+        initiator.is_link_initiator = false;
+        link.link_id = None;
+        link.is_link_initiator = false;
+
```

Comment:
@deveusss we need to add logic to either pause or remove both streams here. I'd prefer the "pause".

The idea is that if there is a bi-directional stream, then either side breaks the link & both streams pause. This way each stream owner must assess if they continue or not.
Think of situation where 2 DAOs decide to do token swap, but then have a disagreement. 1 DAO detachs their side, causing both sides to pause. This makes sure that a bi-directional stream is fair, so that a detach doesnt let another  linked stream continue if it was previously linked.